File: model-manager/load.py
import json, os, sys, time, urllib.request

def env_req(name):
    v = os.environ.get(name)
    if not v:
        print(f"[fatal] missing env: {name}", file=sys.stderr)
        sys.exit(2)
    return v

def parse_duration(text):
    s = text.strip().lower()
    try:
        return float(s)
    except:
        units = {"s":1,"m":60,"h":3600}
        if len(s) >= 2 and s[-1] in units:
            try:
                return float(s[:-1]) * units[s[-1]]
            except:
                pass
    print(f"[fatal] invalid duration: {text}", file=sys.stderr)
    sys.exit(2)

def http_get_json(url, timeout):
    with urllib.request.urlopen(url, timeout=timeout) as r:
        if r.status//100!=2: raise RuntimeError(f"GET {url} {r.status}")
        return json.loads(r.read().decode())

def http_post_json(url, payload, timeout):
    data=json.dumps(payload).encode()
    req=urllib.request.Request(url,data=data,headers={"Content-Type":"application/json"})
    with urllib.request.urlopen(req,timeout=timeout) as r:
        if r.status//100!=2: raise RuntimeError(f"POST {url} {r.status}")
        txt=r.read().decode()
        return json.loads(txt) if txt else {}

def wait_for_api(base_url, startup_timeout, request_timeout):
    deadline=time.time()+startup_timeout; delay=0.5
    while True:
        try:
            http_get_json(f"{base_url}/api/tags",request_timeout); return
        except:
            if time.time()>=deadline: raise
            time.sleep(delay); delay=min(delay*1.5,5.0)

def list_tags(base_url,timeout): return http_get_json(f"{base_url}/api/tags",timeout)

def have_model(tags,name): return any(e.get("model")==name for e in tags.get("models",[]))

def ensure_model(base_url,name,timeout):
    http_post_json(f"{base_url}/api/pull",{"model":name,"stream":False},timeout)
    if not have_model(list_tags(base_url,timeout),name): raise RuntimeError(f"pull failed {name}")

def reconcile_once(base_url,models,timeout):
    if not models: print("[fatal] REQUIRED_MODELS is empty",file=sys.stderr); return -1
    tags=list_tags(base_url,timeout); changed=0
    for m in models:
        if have_model(tags,m): print(f"[ok] {m}"); continue
        print(f"[pull] {m}"); ensure_model(base_url,m,timeout); print(f"[done] {m}"); changed+=1
        tags=list_tags(base_url,timeout)
    return changed

def main():
    base = env_req("OLLAMA_BASE_URL")
    required_raw = env_req("REQUIRED_MODELS")
    startup = parse_duration(env_req("STARTUP_TIMEOUT"))
    req_timeout = parse_duration(env_req("REQUEST_TIMEOUT"))
    loop = parse_duration(env_req("LOOP_INTERVAL"))
    required = [m for m in required_raw.split() if m]
    wait_for_api(base,startup,req_timeout)
    if loop<=0:
        return 0 if reconcile_once(base,required,req_timeout) >= 0 else 1
    while True:
        try: reconcile_once(base,required,req_timeout)
        except Exception as e: print(f"[error] {e}",file=sys.stderr)
        time.sleep(loop)

if __name__=="__main__":
    sys.exit(main())

File: model-manager/load.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"
)

type tagsResponse struct {
	Models []struct {
		Name string `json:"name"`
	} `json:"models"`
}

type pullEvent struct {
	Status    string `json:"status"`
	Error     string `json:"error"`
	Digest    string `json:"digest"`
	Total     int64  `json:"total"`
	Completed int64  `json:"completed"`
}

func requireEnv(key string) (string, error) {
	v := strings.TrimSpace(os.Getenv(key))
	if v == "" {
		return "", fmt.Errorf("required env %s not set", key)
	}
	return v, nil
}

func parseModels(raw string) ([]string, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return nil, errors.New("REQUIRED_MODELS must not be empty")
	}
	fields := strings.Fields(strings.ReplaceAll(raw, ",", " "))
	var out []string
	for _, f := range fields {
		f = strings.TrimSpace(f)
		if f != "" {
			out = append(out, f)
		}
	}
	if len(out) == 0 {
		return nil, errors.New("REQUIRED_MODELS parsed to zero entries")
	}
	return out, nil
}

func mustParseDuration(val, key string) (time.Duration, error) {
	d, err := time.ParseDuration(val)
	if err != nil {
		return 0, fmt.Errorf("%s invalid duration %q: %w", key, val, err)
	}
	return d, nil
}

func httpClient(timeout time.Duration) *http.Client {
	tr := &http.Transport{
		DialContext:           (&net.Dialer{Timeout: timeout}).DialContext,
		TLSHandshakeTimeout:   timeout,
		ResponseHeaderTimeout: timeout,
		ExpectContinueTimeout: timeout,
	}
	return &http.Client{Transport: tr, Timeout: timeout}
}

func pingOllama(ctx context.Context, base string, timeout time.Duration) error {
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, base+"/api/tags", nil)
	resp, err := httpClient(timeout).Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fmt.Errorf("ollama /api/tags status %d", resp.StatusCode)
	}
	return nil
}

func listModels(ctx context.Context, base string, timeout time.Duration) (map[string]struct{}, error) {
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, base+"/api/tags", nil)
	resp, err := httpClient(timeout).Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("status %d", resp.StatusCode)
	}
	var tr tagsResponse
	if err := json.NewDecoder(resp.Body).Decode(&tr); err != nil {
		return nil, err
	}
	have := make(map[string]struct{}, len(tr.Models))
	for _, m := range tr.Models {
		have[m.Name] = struct{}{}
	}
	return have, nil
}

func pullModel(ctx context.Context, base string, timeout time.Duration, model string) error {
	body := strings.NewReader(`{"name":"` + model + `"}`)
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, base+"/api/pull", body)
	req.Header.Set("Content-Type", "application/json")
	resp, err := httpClient(timeout).Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 8192))
		return fmt.Errorf("pull %s failed: status %d: %s", model, resp.StatusCode, string(b))
	}
	dec := json.NewDecoder(bufio.NewReader(resp.Body))
	for {
		var ev pullEvent
		if err := dec.Decode(&ev); err != nil {
			if err == io.EOF {
				break
			}
			return fmt.Errorf("pull %s: decode: %w", model, err)
		}
		if ev.Error != "" {
			return fmt.Errorf("pull %s: %s", model, ev.Error)
		}
		if ev.Status != "" {
			log.Printf("pull %s: %s", model, ev.Status)
		}
		if ev.Status == "success" {
			return nil
		}
	}
	return fmt.Errorf("pull %s: stream ended without success", model)
}

func ensureAll(ctx context.Context, base string, timeout time.Duration, required []string) error {
	have, err := listModels(ctx, base, timeout)
	if err != nil {
		return err
	}
	var missing []string
	for _, need := range required {
		if _, ok := have[need]; !ok {
			missing = append(missing, need)
		}
	}
	for _, m := range missing {
		log.Printf("pulling missing model: %s", m)
		if err := pullModel(ctx, base, timeout, m); err != nil {
			return err
		}
	}
	return nil
}

func waitUntilReady(ctx context.Context, base string, reqTimeout, startupLimit time.Duration, required []string) error {
	deadline := time.Now().Add(startupLimit)
	var last error
	for {
		if time.Now().After(deadline) {
			if last == nil {
				last = errors.New("startup timeout")
			}
			return last
		}
		if err := pingOllama(ctx, base, reqTimeout); err != nil {
			last = fmt.Errorf("ollama not responding: %w", err)
			time.Sleep(2 * time.Second)
			continue
		}
		if err := ensureAll(ctx, base, reqTimeout, required); err != nil {
			last = fmt.Errorf("ensure models: %w", err)
			time.Sleep(2 * time.Second)
			continue
		}
		return nil
	}
}

func main() {
	ollama, err := requireEnv("OLLAMA_BASE_URL")
	if err != nil {
		log.Fatal(err)
	}
	rawModels, err := requireEnv("REQUIRED_MODELS")
	if err != nil {
		log.Fatal(err)
	}
	required, err := parseModels(rawModels)
	if err != nil {
		log.Fatal(err)
	}
	reqTimeoutStr, err := requireEnv("REQUEST_TIMEOUT")
	if err != nil {
		log.Fatal(err)
	}
	startupTimeoutStr, err := requireEnv("STARTUP_TIMEOUT")
	if err != nil {
		log.Fatal(err)
	}
	reqTimeout, err := mustParseDuration(reqTimeoutStr, "REQUEST_TIMEOUT")
	if err != nil {
		log.Fatal(err)
	}
	startupLimit, err := mustParseDuration(startupTimeoutStr, "STARTUP_TIMEOUT")
	if err != nil {
		log.Fatal(err)
	}
	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
	defer stop()
	if err := waitUntilReady(ctx, ollama, reqTimeout, startupLimit, required); err != nil {
		log.Fatalf("startup failed: %v", err)
	}
	log.Printf("all required models present")
	<-ctx.Done()
}

